btree/btree.c

```diff
@@ -0,0 +1,13037 @@
+			      struct segaddr *addr, int nxt);
+static void       bnode_put  (struct node_op *op, struct nd *node);
+
+static void bnode_crc_validate(struct nd *node);
+
+static int64_t    bnode_free(struct node_op *op, struct nd *node,
+			     struct m0_be_tx *tx, int nxt);
+static int64_t    bnode_alloc(struct node_op *op, struct td *tree, int shift,
+			      const struct node_type *nt,
+			      const enum m0_btree_crc_type crc_type,
+			      int ksize, int vsize,
+			      struct m0_be_tx *tx, int nxt);
+static void bnode_op_fini(struct node_op *op);
+static int bnode_access(struct segaddr *addr, int nxt);
+static int  bnode_init(struct segaddr *addr, int ksize, int vsize, int nsize,
+		       const struct node_type *nt,
+		       const enum m0_btree_crc_type crc_type,uint64_t gen,
+		       struct m0_fid fid, int nxt);
+static uint32_t bnode_crctype_get(const struct nd *node);
+/* Returns the number of valid key in the node. */
+static int  bnode_count(const struct nd *node);
+
+/* Returns the number of records in the node. */
+static int  bnode_count_rec(const struct nd *node);
+static int  bnode_space(const struct nd *node);
+static int  bnode_level(const struct nd *node);
+static int  bnode_nsize(const struct nd *node);
+static int  bnode_keysize(const struct nd *node);
+static int  bnode_valsize(const struct nd *node);
+static bool  bnode_isunderflow(const struct nd *node, bool predict);
+static bool  bnode_isoverflow(const struct nd *node, int max_ksize,
+			      const struct m0_btree_rec *rec);
+
+static void bnode_rec  (struct slot *slot);
+static void bnode_key  (struct slot *slot);
+static void bnode_child(struct slot *slot, struct segaddr *addr);
+static bool bnode_isfit(struct slot *slot);
+static void bnode_done(struct slot *slot, bool modified);
+static void bnode_make(struct slot *slot);
+static void bnode_val_resize(struct slot *slot, int vsize_diff);
+
+static bool bnode_find (struct slot *slot, struct m0_btree_key *key);
+static void bnode_seq_cnt_update (struct nd *node);
+static void bnode_fix  (const struct nd *node);
+static void bnode_del  (const struct nd *node, int idx);
+static void bnode_set_level  (const struct nd *node, uint8_t new_level);
+static void bnode_set_rec_count(const struct nd *node, uint16_t count);
+static void bnode_move (struct nd *src, struct nd *tgt, enum direction dir,
+		        int nr);
+
+static void bnode_capture(struct slot *slot, struct m0_be_tx *tx);
+
+static void bnode_lock(struct nd *node);
+static void bnode_unlock(struct nd *node);
+static void bnode_fini(const struct nd *node);
+
+/**
+ * Common node header.
+ *
+ * This structure is located at the beginning of every node, right after
+ * m0_format_header. It is used by the segment operations (node_op) to identify
+ * node and tree types.
+ */
+struct node_header {
+	uint32_t      h_node_type;
+	uint32_t      h_tree_type;
+	uint32_t      h_crc_type;
+	uint64_t      h_gen;
```

Comment:
If h_gen is keeping the version info then it can be the first element

Reply:
h_gen keeps the segment generation id, and does not contain version info, so no need to make it first element.